browser/brave_rewards/rewards_tab_helper.cc

```diff
@@ -22,85 +32,103 @@
 
-#if BUILDFLAG(ENABLE_IPFS)
-#include "brave/components/ipfs/ipfs_constants.h"
-#endif
-
-using blink::mojom::ResourceType;
-
-// DEFINE_WEB_CONTENTS_USER_DATA_KEY(brave_rewards::RewardsTabHelper);
-
 namespace brave_rewards {
 
 RewardsTabHelper::RewardsTabHelper(content::WebContents* web_contents)
-    : WebContentsObserver(web_contents),
-      content::WebContentsUserData<RewardsTabHelper>(*web_contents),
+    : content::WebContentsUserData<RewardsTabHelper>(*web_contents),
+      WebContentsObserver(web_contents),
       tab_id_(sessions::SessionTabHelper::IdForTab(web_contents)) {
-  if (!tab_id_.is_valid())
-    return;
+  if (tab_id_.is_valid()) {
+    rewards_service_ = RewardsServiceFactory::GetForProfile(
+        Profile::FromBrowserContext(GetWebContents().GetBrowserContext()));
+  }
+
+  if (rewards_service_) {
+    rewards_service_->AddObserver(this);
+  }
 
 #if !BUILDFLAG(IS_ANDROID)
   BrowserList::AddObserver(this);
 #endif
-  Profile* profile =
-      Profile::FromBrowserContext(web_contents->GetBrowserContext());
-  rewards_service_ = RewardsServiceFactory::GetForProfile(profile);
-  if (rewards_service_)
-    rewards_service_->AddObserver(this);
 }
 
 RewardsTabHelper::~RewardsTabHelper() {
-  if (rewards_service_)
+  if (rewards_service_) {
     rewards_service_->RemoveObserver(this);
+  }
 #if !BUILDFLAG(IS_ANDROID)
   BrowserList::RemoveObserver(this);
 #endif
 }
 
+void RewardsTabHelper::AddObserver(Observer* observer) {
+  observer_list_.AddObserver(observer);
+}
+
+void RewardsTabHelper::RemoveObserver(Observer* observer) {
+  observer_list_.RemoveObserver(observer);
+}
+
+void RewardsTabHelper::SetPublisherIdForTab(const std::string& publisher_id) {
+  if (publisher_id != publisher_id_) {
+    publisher_id_ = publisher_id;
+    for (auto& observer : observer_list_) {
+      observer.OnPublisherForTabUpdated(publisher_id_);
+    }
+  }
+}
+
 void RewardsTabHelper::DidFinishLoad(
     content::RenderFrameHost* render_frame_host,
     const GURL& validated_url) {
-  if (!rewards_service_ || render_frame_host->GetParent())
+  if (!rewards_service_ || render_frame_host->GetParent()) {
     return;
+  }
 
 #if BUILDFLAG(ENABLE_IPFS)
-  auto ipns_url = web_contents()->GetURL();
+  auto ipns_url = GetWebContents().GetLastCommittedURL();
   if (ipns_url.SchemeIs(ipfs::kIPNSScheme)) {
     rewards_service_->OnLoad(tab_id_, ipns_url);
     return;
   }
 #endif
+
   rewards_service_->OnLoad(tab_id_, validated_url);
 }
 
 void RewardsTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
-  if (!rewards_service_ || !handle->IsInMainFrame() ||
-      !handle->HasCommitted() || handle->IsDownload())
+  if (!handle->IsInMainFrame() || !handle->HasCommitted() ||
+      handle->IsDownload()) {
     return;
+  }
+
+  auto id = GetPublisherIdFromURL(GetWebContents().GetLastCommittedURL());
+  SetPublisherIdForTab(id ? *id : "");
+
+  MaybeSavePublisherInfo();
```

Comment:
not sure I understand what's happening here

Reply:
I added some additional explanation to the comment in `MaybeSavePublisherInfo`.